File: timer_manager.rb
# timer_manager.rb
require 'io/console'
require 'json'
require 'time'

class Timer
  attr_accessor :name, :duration, :elapsed, :running, :finished

  def initialize(name, duration, elapsed=0, running=false, finished=false)
    @name = name
    @duration = duration
    @elapsed = elapsed
    @running = running
    @finished = finished
  end

  def remaining
    finished ? 0 : [duration - elapsed, 0].max
  end

  def update(dt)
    if running && !finished
      @elapsed += dt
      if @elapsed >= duration
        @elapsed = duration
        @running = false
        @finished = true
        return true
      end
    end
    false
  end

  def to_h
    {name: @name, duration: @duration, elapsed: @elapsed, running: @running, finished: @finished}
  end

  def self.from_h(h)
    new(h['name'], h['duration'], h['elapsed'], h['running'], h['finished'])
  end
end

class TimerManager
  def initialize(save_file='timers.json')
    @save_file = save_file
    @timers = []
    @selected = 0
    @running = true
    @last_update = Time.now
    load
    if @timers.empty?
      @timers << Timer.new("Pomodoro", 25*60)
      @timers << Timer.new("Break", 5*60)
    end
  end

  def add_timer(name, duration)
    @timers << Timer.new(name, duration)
    save
  end

  def remove_timer(idx)
    return if idx < 0 || idx >= @timers.size
    @timers.delete_at(idx)
    @selected = @timers.size-1 if @selected >= @timers.size
    save
  end

  def toggle(idx)
    return if idx < 0 || idx >= @timers.size
    t = @timers[idx]
    if t.finished
      t.elapsed = 0
      t.finished = false
      t.running = true
    else
      t.running = !t.running
    end
    save
  end

  def reset(idx)
    return if idx < 0 || idx >= @timers.size
    t = @timers[idx]
    t.elapsed = 0
    t.running = false
    t.finished = false
    save
  end

  def update
    now = Time.now
    dt = now - @last_update
    @last_update = now
    @timers.each do |t|
      if t.update(dt)
        puts "\n!!! Timer '#{t.name}' finished !!!"
      end
    end
  end

  def save
    File.write(@save_file, JSON.pretty_generate(@timers.map(&:to_h)))
  end

  def load
    if File.exist?(@save_file)
      begin
        data = JSON.parse(File.read(@save_file))
        @timers = data.map { |h| Timer.from_h(h) }
      rescue
        @timers = []
      end
    end
  end

  def run
    system('stty -echo -icanon min 1') rescue nil
    at_exit { system('stty echo icanon'); print "\e[?25h" }
    print "\e[?25l"

    loop do
      print "\e[2J\e[H"
      puts "=== Timer Manager ==="
      puts "Timers: #{@timers.size}   [a]dd  [d]elete  [Space]toggle  [r]eset  [q]uit"

      @timers.each_with_index do |t, i|
        rem = t.remaining
        status = t.finished ? " ✓" : (t.running ? " ▶" : " ⏸")
        line = sprintf("%d. %s %s  %02.0f:%02.0f / %02.0f:%02.0f",
                       i+1, t.name, status, rem/60, rem%60, t.duration/60, t.duration%60)
        if i == @selected
          print "\e[7m"
        else
          color = t.running ? "\e[32m" : (t.finished ? "\e[31m" : "\e[33m")
          print color
        end
        puts line
        print "\e[0m"
      end
      puts "\nCommands: a=add, d=delete, Space=toggle, r=reset, q=quit"

      # Non-blocking input
      begin
        c = STDIN.getc
        case c
        when 'q' then break
        when 'a'
          print "Enter name: "
          name = STDIN.gets.chomp
          print "Enter duration (seconds): "
          dur = STDIN.gets.chomp.to_f
          add_timer(name, dur)
        when 'd'
          remove_timer(@selected) if @timers.any?
        when ' '
          toggle(@selected) if @timers.any?
        when 'r'
          reset(@selected) if @timers.any?
        when "\e"
          c2 = STDIN.getc
          if c2 == '['
            c3 = STDIN.getc
            case c3
            when 'A' then @selected = [0, @selected-1].max
            when 'B' then @selected = [@timers.size-1, @selected+1].min
            end
          end
        end
      rescue
        # no input
      end

      update
      sleep 0.05
    end
    save
    print "\e[?25h"
  end
end

if __FILE__ == $0
  manager = TimerManager.new
  manager.run
end

File: timer_manager.go
// timer_manager.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Timer struct {
	Name     string  `json:"name"`
	Duration float64 `json:"duration"` // seconds
	Elapsed  float64 `json:"elapsed"`
	Running  bool    `json:"running"`
	Finished bool    `json:"finished"`
}

func (t *Timer) Remaining() float64 {
	if t.Finished {
		return 0
	}
	rem := t.Duration - t.Elapsed
	if rem < 0 {
		return 0
	}
	return rem
}

func (t *Timer) Update(dt float64) bool {
	if t.Running && !t.Finished {
		t.Elapsed += dt
		if t.Elapsed >= t.Duration {
			t.Elapsed = t.Duration
			t.Running = false
			t.Finished = true
			return true
		}
	}
	return false
}

type TimerManager struct {
	Timers   []Timer
	Selected int
	SaveFile string
	running  bool
}

func NewTimerManager(saveFile string) *TimerManager {
	tm := &TimerManager{
		Timers:   []Timer{},
		Selected: 0,
		SaveFile: saveFile,
		running:  true,
	}
	tm.Load()
	if len(tm.Timers) == 0 {
		tm.Timers = append(tm.Timers, Timer{Name: "Pomodoro", Duration: 25 * 60, Elapsed: 0, Running: false, Finished: false})
		tm.Timers = append(tm.Timers, Timer{Name: "Break", Duration: 5 * 60, Elapsed: 0, Running: false, Finished: false})
	}
	return tm
}

func (tm *TimerManager) AddTimer(name string, duration float64) {
	tm.Timers = append(tm.Timers, Timer{Name: name, Duration: duration, Elapsed: 0, Running: false, Finished: false})
	tm.Save()
}

func (tm *TimerManager) RemoveTimer(idx int) {
	if idx < 0 || idx >= len(tm.Timers) {
		return
	}
	tm.Timers = append(tm.Timers[:idx], tm.Timers[idx+1:]...)
	if tm.Selected >= len(tm.Timers) {
		tm.Selected = len(tm.Timers) - 1
	}
	tm.Save()
}

func (tm *TimerManager) Toggle(idx int) {
	if idx < 0 || idx >= len(tm.Timers) {
		return
	}
	t := &tm.Timers[idx]
	if t.Finished {
		t.Elapsed = 0
		t.Finished = false
		t.Running = true
	} else {
		t.Running = !t.Running
	}
	tm.Save()
}

func (tm *TimerManager) Reset(idx int) {
	if idx < 0 || idx >= len(tm.Timers) {
		return
	}
	t := &tm.Timers[idx]
	t.Elapsed = 0
	t.Running = false
	t.Finished = false
	tm.Save()
}

func (tm *TimerManager) Update(dt float64) {
	for i := range tm.Timers {
		if tm.Timers[i].Update(dt) {
			fmt.Printf("\n!!! Timer '%s' finished !!!\n", tm.Timers[i].Name)
		}
	}
}

func (tm *TimerManager) Save() {
	data, err := json.MarshalIndent(tm.Timers, "", "  ")
	if err == nil {
		os.WriteFile(tm.SaveFile, data, 0644)
	}
}

func (tm *TimerManager) Load() {
	data, err := os.ReadFile(tm.SaveFile)
	if err == nil {
		json.Unmarshal(data, &tm.Timers)
	}
}

func (tm *TimerManager) Run() {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Print("\033[?25l") // hide cursor
	defer fmt.Print("\033[?25h")
	lastUpdate := time.Now()

	for tm.running {
		// Clear screen
		fmt.Print("\033[2J\033[H")
		fmt.Println("=== Timer Manager ===")
		fmt.Printf("Timers: %d   [a]dd  [d]elete  [Space]toggle  [r]eset  [q]uit\n", len(tm.Timers))

		// Display timers
		for i, t := range tm.Timers {
			remaining := t.Remaining()
			status := " ▶"
			if t.Finished {
				status = " ✓"
			} else if !t.Running {
				status = " ⏸"
			}
			line := fmt.Sprintf("%d. %s %s  %02.0f:%02.0f / %02.0f:%02.0f",
				i+1, t.Name, status,
				remaining/60, remaining,
				t.Duration/60, t.Duration)
			if i == tm.Selected {
				fmt.Print("\033[7m") // reverse
			}
			fmt.Println(line)
			fmt.Print("\033[0m")
		}
		fmt.Println("\nCommands: a=add, d=delete, Space=toggle, r=reset, q=quit")

		// Input (non-blocking)
		scanner.Scan()
		cmd := scanner.Text()
		switch cmd {
		case "q":
			tm.running = false
		case "a":
			fmt.Print("Enter name: ")
			scanner.Scan()
			name := scanner.Text()
			fmt.Print("Enter duration (seconds): ")
			scanner.Scan()
			durStr := scanner.Text()
			dur, err := strconv.ParseFloat(durStr, 64)
			if err == nil {
				tm.AddTimer(name, dur)
			}
		case "d":
			if len(tm.Timers) > 0 {
				tm.RemoveTimer(tm.Selected)
			}
		case " ":
			if len(tm.Timers) > 0 {
				tm.Toggle(tm.Selected)
			}
		case "r":
			if len(tm.Timers) > 0 {
				tm.Reset(tm.Selected)
			}
		case "\x1b[A": // up
			if tm.Selected > 0 {
				tm.Selected--
			}
		case "\x1b[B": // down
			if tm.Selected < len(tm.Timers)-1 {
				tm.Selected++
			}
		default:
			// ignore
		}

		// Update timers
		now := time.Now()
		dt := now.Sub(lastUpdate).Seconds()
		lastUpdate = now
		tm.Update(dt)
		time.Sleep(50 * time.Millisecond)
	}
	tm.Save()
}

func main() {
	manager := NewTimerManager("timers.json")
	manager.Run()
}
